File: StronglyTypedEnumConverter_VbTests/CowboyType.vb


Imports System.Collections.Generic
Imports System.Linq

Namespace StronglyTypedEnumConverter
    Class CowboyType
        Private Sub New()
        End Sub

        Public Shared ReadOnly Good As New CowboyType()
        Public Shared ReadOnly Bad As New CowboyType()
        Public Shared ReadOnly Ugly As New CowboyType()

        Public Shared Iterator Function All() As IEnumerable(Of CowboyType)
            Yield Good
            Yield Bad
            Yield Ugly
        End Function

        Public Overrides Function ToString() As String
            Dim map = New Dictionary(Of CowboyType, String)() From { _
                {Good, "Good"}, _
                {Bad, "Bad"}, _
                {Ugly, "Ugly"} _
            }

            Return map(Me)
        End Function

        Public Shared Function FromString(value As String) As CowboyType
            If value Is Nothing Then
                Throw New ArgumentNullException("value")
            End If

            Dim result = All().FirstOrDefault(Function(x) x.ToString() = value)
            If result IsNot Nothing Then
                Return result
            End If

            Throw New ArgumentOutOfRangeException("value", value, "Invalid CowboyType")
        End Function

        Public Shared Narrowing Operator CType(value As CowboyType) As Integer
            Dim map = New Dictionary(Of CowboyType, Integer)() From { _
                {Good, 0}, _
                {Bad, 1}, _
                {Ugly, 2} _
            }

            Return map(value)
        End Operator

        Public Shared Narrowing Operator CType(value As Integer) As CowboyType
            Dim result = All().FirstOrDefault(Function(x) CInt(x) = value)
            If result IsNot Nothing Then
                Return result
            End If

            Throw New InvalidCastException("The value " & value & " is not a valid CowboyType")
        End Operator

    End Class
End Namespace

File: StronglyTypedEnumConverter_Tests/CowboyType.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Reflection;

class CowboyType : IComparable<CowboyType>
{

    private CowboyType(string name, int value)
    {
        _name = name;
        _value = value;
    }

    #region Members

    public static readonly CowboyType Good = new CowboyType(nameof(Good), 0);
    public static readonly CowboyType Bad = new CowboyType(nameof(Bad), 1);
    public static readonly CowboyType Ugly = new CowboyType(nameof(Ugly), 2);

    #endregion


    #region All

    public static IEnumerable<CowboyType> All() => All<CowboyType>();

    private static IEnumerable<T> All<T>()
    {
        var type = typeof(T);
        return type.GetFields(BindingFlags.Public | BindingFlags.Static)
            .Where(field => field.IsInitOnly)
            .Where(field => field.FieldType == type)
            .Select(field => field.GetValue(null))
            .Cast<T>();
    }

    #endregion


    #region To/From String

    private readonly string _name;

    public override string ToString() => _name;

    public static CowboyType FromString(string value)
    {
        if (value == null) throw new ArgumentNullException(nameof(value));

        var result = All().FirstOrDefault(x => x.ToString() == value);
        if (result != null) return result;

        throw new ArgumentOutOfRangeException(nameof(value), value, "Invalid CowboyType");
    }

    #endregion


    #region Cast to/from Underlying Type

    private readonly int _value;
    public static explicit operator int(CowboyType value) => value._value;

    public static explicit operator CowboyType(int value)
    {
        var result = All().FirstOrDefault(x => (int)x == value);
        if (result != null) return result;

        throw new InvalidCastException("The value " + value + " is not a valid CowboyType");
    }

    #endregion

    #region IComparable

    public int CompareTo(CowboyType other)
    {
        var results = new[]
        {
            ((int) this).CompareTo((int) other)
        };
        return results
            .SkipWhile(diff => diff == 0)
            .FirstOrDefault();
    }

    public static bool operator <(CowboyType lhs, CowboyType rhs) => lhs.CompareTo(rhs) < 0;

    public static bool operator >(CowboyType lhs, CowboyType rhs) => lhs.CompareTo(rhs) > 0;

    public static bool operator <=(CowboyType lhs, CowboyType rhs) => lhs.CompareTo(rhs) <= 0;
    public static bool operator >=(CowboyType lhs, CowboyType rhs) => lhs.CompareTo(rhs) >= 0;

    #endregion

}
